MarkBind/markbind
# Fix `puml` not syncing Resolve #1617

**What is the purpose of this pull request?**




**Overview of changes:**
Resolve #1617
By introducing a lock service to wait for asynchronous functions to complete:
```
// Importing LockManager module from the specified directory
const LockManager = require('../../utils/LockManager');

// Start an async function
const lockId = LockManager.createLock();

// Stop of an async function
LockManager.deleteLock(lockId);

// Wait for all async functions to complete
await LockManager.waitForLockRelease();
```

Fix the issue where all distinct PUML diagrams were being reprocessed on every hot reload by introducing a hashmap tracker to keep track of the diagrams.
- The hash key comes from the content and the file path of the PUML diagram.

**Anything you'd like to highlight/discuss:**
Performance Analysis:

The implementation of the lock service has minimal impact on PUML processing performance. 

Testing was conducted on 70 PUML diagrams:

- Original Configuration (without lock service)
    `- o`
    <img width="642" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/23dec5c5-3fab-48c8-8f83-264f9ee267a0">
    - CPU Utilization: 12% to 93%
    - RAM Usage: 60% to 68%
    - Hot Reload CPU Utilization: 12% to 94%
    - Hot Reload RAM Usage: 60% to 69%
  `- d`
    <img width="698" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/e4b89b61-6e55-4f18-9adb-25793b30ae75">

    - Initial CPU Utilization: 12% to 97%
    - Initial RAM Usage: 63% to 78%
    - Hot Reload CPU Utilization: 6% to 94%
    - Hot Reload RAM Usage: 63% to 75%


- Singleton Implementation
   `-o`
    <img width="665" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/265be980-f879-43e6-a8dc-305323df74af">
  Initial Performance Statistics:
    - CPU Utilization: 12% to 95%
    - RAM Usage: 63% to 72%
    - On every single diagram hot reload: RAM shows no change, CPU increases by 6%

   `-d`
   <img width="470" alt="image" src="https://github.com/MarkBind/markbind/assets/30100720/330f3d77-2329-454b-9804-7d6c29aab3ae">

  Initial Performance Statistics:
  - CPU Utilization: 7% to 96%
  - RAM Usage: 63% to 74%
  - On every single diagram hot reload: RAM shows no change, CPU increases by 6%


These results indicate that the lock service does not significantly affect PUML processing performance.
In fact, it greatly reduces the CPU computation on every hot reload, leading to a more efficient processing system.


**Testing instructions:**
1. Check out the PR locally.
2. Create a new MarkBind project using `markbind init`.
3. Run `markbind serve -o`.
4. Experiment with the sample:

```markdown
<puml width="300">
@startuml
alice -> bob ++ : hellos
bob -> bob ++ : self call
bob -> bib ++  #005500 : a
bob -> george ** : create
return done
return rc
bob -> george !! : delete
return success
@enduml
</puml>
```
5. Edit the sample and wait for the hot reload to see the effect.



**Proposed commit message: (wrap lines at 72 characters)**
Introduce LockManager for Async Handling and Optimize PUML Diagrams

Introduce the LockManager module to manage the lifecycle of asynchronous
functions, ensuring robust and consistent handling within synchronous
methods.

Utilize the createLock, deleteLock, and waitForLockRelease methods for
starting, stopping, and waiting for async functions respectively.

Code snippet:

```javascript
const LockManager = require('path_to/LockManager');
const lockId = LockManager.createLock();
LockManager.deleteLock(lockId);
await LockManager.waitForLockRelease();
```

Fix the issue where all distinct PUML diagrams were being reprocessed on
every hot reload by introducing a hashmap tracker.

- The hash key comes from the content and the file name of the PUML diagram.
- The original \`processedDiagrams.clear()\` has been updated to
  \`cleanDiagrams(processedDiagrams)\`.

By implementing these changes, the code is made more maintainable,
asynchronous functions are properly managed, and PUML diagram processing
is optimized.



---

**Checklist:** :ballot_box_with_check:
